sql/catalyst/src/main/scala/org/apache/spark/sql/Encoders.scala

```diff
@@ -209,7 +209,12 @@ object Encoders {
           BoundReference(0, ObjectType(classOf[AnyRef]), nullable = true), kryo = useKryo)),
       deserializer =
         DecodeUsingSerializer[T](
-          GetColumnByOrdinal(0, BinaryType), classTag[T], kryo = useKryo),
+          UpCast(
```

Comment:
I think we can just use `Cast`. `Upcast` is mainly for some numeric check, e.g. long should not be casted to int.